src/stream/src/executor/dedup/append_only_dedup.rs

```diff
@@ -0,0 +1,295 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use futures::{stream, StreamExt};
+use futures_async_stream::try_stream;
+use itertools::Itertools;
+use risingwave_common::array::{Op, StreamChunk};
+use risingwave_common::buffer::BitmapBuilder;
+use risingwave_common::catalog::Schema;
+use risingwave_common::row::{OwnedRow, Row, RowExt};
+use risingwave_storage::StateStore;
+
+use super::cache::DedupCache;
+use crate::common::table::state_table::StateTable;
+use crate::executor::error::StreamExecutorError;
+use crate::executor::{
+    expect_first_barrier, ActorContextRef, BoxedExecutor, BoxedMessageStream, Executor, Message,
+    PkIndices, PkIndicesRef, StreamExecutorResult,
+};
+use crate::task::AtomicU64Ref;
+
+/// [`AppendOnlyDedupExecutor`] drops any message that has duplicate pk columns with previous
+/// messages. It only accepts append-only input, and its output will be append-only as well.
+pub struct AppendOnlyDedupExecutor<S: StateStore> {
+    input: Option<BoxedExecutor>,
+    state_table: StateTable<S>,
+    cache: DedupCache<OwnedRow>,
+
+    pk_indices: PkIndices,
+    identity: String,
+    schema: Schema,
+    ctx: ActorContextRef,
+}
+
+impl<S: StateStore> AppendOnlyDedupExecutor<S> {
+    pub fn new(
+        input: BoxedExecutor,
+        state_table: StateTable<S>,
+        pk_indices: PkIndices,
+        executor_id: u64,
+        ctx: ActorContextRef,
+        watermark_epoch: AtomicU64Ref,
+    ) -> Self {
+        let schema = input.schema().clone();
+        Self {
+            input: Some(input),
+            state_table,
+            cache: DedupCache::new(watermark_epoch),
+            pk_indices,
+            identity: format!("AppendOnlyDedupExecutor {:X}", executor_id),
+            schema,
+            ctx,
+        }
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    async fn executor_inner(mut self) {
+        let mut input = self.input.take().unwrap().execute();
+
+        // Consume the first barrier message and initialize state table.
+        let barrier = expect_first_barrier(&mut input).await?;
+        self.state_table.init_epoch(barrier.epoch);
+
+        // The first barrier message should be propagated.
+        yield Message::Barrier(barrier);
+
+        #[for_await]
+        for msg in input {
+            match msg? {
+                Message::Chunk(chunk) => {
+                    // Append-only dedup executor only receives INSERT messages.
+                    debug_assert!(chunk.ops().iter().all(|&op| op == Op::Insert));
+
+                    // Extract pk for all rows (regardless of visibility) in the chunk.
+                    let keys = chunk
+                        .data_chunk()
+                        .rows_with_holes()
+                        .map(|row_ref| {
+                            row_ref.map(|row| row.project(self.pk_indices()).to_owned_row())
```

Comment:
Is it possible to make `populate_cache` directly take `impl Row`, so that we don't need to allocate here?

Reply:
We'll `dedup_insert` the key into the cache later, so it seems inevitable to make `key` an owned value.

https://github.com/risingwavelabs/risingwave/blob/6c27de11b21591f948a0c6a8a489806bad689a8c/src/stream/src/executor/dedup/append_only_dedup.rs#L100-L103